contracts/BonusRewards.sol

```diff
@@ -3,11 +3,363 @@
+    require(pools[_token].lastUpdatedAt == 0, "BonusRewards: lpToken, not allowed");
+
+    if (_token == address(0)) { // token address(0) = ETH
+      payable(owner()).transfer(address(this).balance);
+    } else {
+      uint256 balance = IERC20(_token).balanceOf(address(this));
+      if (bonusTokenAddrMap[_token] == 1) {
+        // bonus token
+        Bonus memory bonus = pools[_lpToken].bonuses[_poolBonusId];
+        require(bonus.bonusTokenAddr == _token, "BonusRewards: wrong pool");
+        require(bonus.endTime + WEEK < block.timestamp, "BonusRewards: not ready");
+        balance = bonus.remBonus;
+        pools[_lpToken].bonuses[_poolBonusId].remBonus = 0;
+      }
+
+      IERC20(_token).transfer(owner(), balance);
+    }
+  }
+
+  function setResponders(address[] calldata _responders) external override onlyOwner {
+    responders = _responders;
+  }
+
+  function setPaused(bool _paused) external override {
+    require(_isAuthorized(msg.sender, responders), "BonusRewards: caller not responder");
+    paused = _paused;
+  }
+
+  /// @notice update pool's bonus per staked token till current block timestamp, do nothing if pool does not exist
+  function _updatePool(address _lpToken) private {
+    Pool storage pool = pools[_lpToken];
+    uint256 poolLastUpdatedAt = pool.lastUpdatedAt;
+    if (poolLastUpdatedAt == 0 || block.timestamp <= poolLastUpdatedAt) return;
+    pool.lastUpdatedAt = block.timestamp;
+    uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+    if (lpTotal == 0) return;
+
+    for (uint256 i = 0; i < pool.bonuses.length; i ++) {
+      Bonus storage bonus = pool.bonuses[i];
+      if (poolLastUpdatedAt < bonus.endTime && bonus.startTime < block.timestamp) {
+        uint256 bonusForTime = _calRewardsForTime(bonus, poolLastUpdatedAt);
+        bonus.accRewardsPerToken = bonus.accRewardsPerToken + bonusForTime / lpTotal;
+      }
+    }
+  }
+
+  function _updateUserWriteoffs(address _lpToken) private {
+    Bonus[] memory bonuses = pools[_lpToken].bonuses;
+    User storage user = users[_lpToken][msg.sender];
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      // update writeoff to match current acc rewards per token
+      if (user.rewardsWriteoffs.length == i) {
+        user.rewardsWriteoffs.push(user.amount * bonuses[i].accRewardsPerToken / CAL_MULTIPLIER);
+      } else {
+        user.rewardsWriteoffs[i] = user.amount * bonuses[i].accRewardsPerToken / CAL_MULTIPLIER;
+      }
+    }
+  }
+
+  /// @notice tranfer upto what the contract has
+  function _safeTransfer(address _token, uint256 _amount) private returns (uint256 _transferred) {
+    IERC20 token = IERC20(_token);
+    uint256 balance = token.balanceOf(address(this));
+    if (balance > _amount) {
+      token.safeTransfer(msg.sender, _amount);
+      _transferred = _amount;
+    } else if (balance > 0) {
+      token.safeTransfer(msg.sender, balance);
+      _transferred = balance;
+    }
```

Comment:
```suggestion
      uint256 amountToSend = balance > _amount ? _amount : balance;
      token.safeTransfer(msg.sender, amountToSend);
      _transferred = amountToSend;
```
Can be refactored.

Reply:
Refactor ignores balance == 0. Don't do transfer 0 as it costs 1600 gas,.